### PR TITLE
look up EnumAsInfo indexes in enum_values

### LINKED ISSUE #477: EnumAsInfo panic
### snmp_exporter version: output of `snmp_exporter -version`

version=0.16.1

### What device/snmpwalk OID are you using?

snmp.yml:
```
  - name: ciscoEnvMonSupplyStatusIndex
    oid: 1.3.6.1.4.1.9.9.13.1.5.1.1
    type: gauge
    help: Unique index for the power supply being instrumented - 1.3.6.1.4.1.9.9.13.1.5.1.1
    indexes:
    - labelname: ciscoEnvMonSupplyStatusIndex
      type: gauge
```

### What did you do that produced an error?

Set an `EnumAsInfo`

### What did you see instead?

```
snmp_exporter[1138]: panic: Unknown index type EnumAsInfo
snmp_exporter[1138]: goroutine 1918149 [running]:
snmp_exporter[1138]: main.indexOidsAsString(0xc0008ee4e8, 0x1, 0x3, 0xc0011c54f0, 0xa, 0x0, 0xab9200, 0xc000192a40, 0x0, 0x0, ...)
snmp_exporter[1138]: #011/app/collector.go:662 +0x1870
snmp_exporter[1138]: main.indexesToLabels(0xc0008ee4e8, 0x1, 0x3, 0xc00131dc00, 0xc0006bbaa0, 0x4)
snmp_exporter[1138]: #011/app/collector.go:673 +0x169
snmp_exporter[1138]: main.pduToSamples(0xc0008ee4e8, 0x1, 0x3, 0xc0006bbad0, 0xc00131dc00, 0xc0006bbaa0, 0xaaf060, 0xc000721e60, 0xc000d36a50, 0x1, ...)
snmp_exporter[1138]: #011/app/collector.go:326 +0x77
snmp_exporter[1138]: main.collector.Collect(0xc000d6a025, 0x11, 0xc0001752c0, 0xaaf060, 0xc000721e60, 0xc001252e40)
snmp_exporter[1138]: #011/app/collector.go:244 +0x914
snmp_exporter[1138]: github.com/prometheus/client_golang/prometheus.(*Registry).Gather.func1()
snmp_exporter[1138]: #011/app/vendor/github.com/prometheus/client_golang/prometheus/registry.go:443 +0x19d
snmp_exporter[1138]: created by github.com/prometheus/client_golang/prometheus.(*Registry).Gather
snmp_exporter[1138]: #011/app/vendor/github.com/prometheus/client_golang/prometheus/registry.go:454 +0x57d
```

It turns out I made an incorrect override for `ciscoEnvMonSupplyStatusIndex` as an `EnumAsInfo`. So this is possibly a generator safety check that is needed to avoid allowing enum types for OIDs that don't have `enum_values`.

### PR DESCRIPTION
**Pull request by fredericve**
Fixes #477 

Adds the index enum_values in the generated snmp.yml file and then uses that to look it up during scrape.

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -181,10 +181,11 @@ type Metric struct {
...
 type Index struct {
-	Labelname string `yaml:"labelname"`
-	Type      string `yaml:"type"`
-	FixedSize int    `yaml:"fixed_size,omitempty"`
-	Implied   bool   `yaml:"implied,omitempty"`
+	Labelname  string         `yaml:"labelname"`
+	Type       string         `yaml:"type"`
+	FixedSize  int            `yaml:"fixed_size,omitempty"`
+	Implied    bool           `yaml:"implied,omitempty"`
+    EnumValues map[int]string `yaml:"enum_values,omitempty"`
```
Comment (brian-brazil) [inline config/config.go:188]: You need to do a make format.

File: collector.go [hunk 2]
```diff
@@ -669,6 +669,10 @@ func indexOidsAsString(indexOids []int, typ string, fixedSize int, implied bool)
 			parts[i] = o
 		}
 		return fmt.Sprintf("%02X%02X:%02X%02X:%02X%02X:%02X%02X:%02X%02X:%02X%02X:%02X%02X:%02X%02X", parts...), subOid, indexOids
+	case "EnumAsInfo":
+		subOid, indexOids := splitOid(indexOids, 1)
+		return enumValues[subOid[0]], subOid, indexOids
```
Comment (brian-brazil) [inline collector.go:674]: This should have a fallback in case the value isn't within enumValues.

**Review by brian-brazil — COMMENTED** [review]
This looks about right. It'll need unittests for both the generator and exporter.

File: collector.go (hunk 2 above)
Comment (fredericve) [reply collector.go:674]: Got it. What kind of fallback do you suggest? Maybe just the value of `suboid[0]`?

Comment (brian-brazil) [reply collector.go:674]: Yes, that's what we usually do.

**Comment by fredericve** [comment]
Added collector and generator tests. Can you verify if these are OK? I'm not entirely sure it's correct.

**Comment by brian-brazil** [comment]
Thanks!

**Pull request merged**